### ISSUE TITLE
Mobx support?

### ISSUE THREAD

**Issue by thinkpadder1** [hyperapp/hyperapp#103]
So, we have redux/elm like state management. But, what about [mobx](https://github.com/mobxjs/mobx)?

[Here](https://mobx.js.org/) is their user guide.
[Here](https://mobx.js.org/getting-started.html) is a 10 minute interactive tutorial.

![Mobx picture](https://raw.githubusercontent.com/mobxjs/mobx/master/docs/flow.png)

I used to write with Redux, but Mobx changed everything for me. Besides having data react automatically, the most important thing for me was being able to use classes with real references (no normalizing of data required). Also, Typescript... is amazing (Mobx is written in it).

**Comment by jbucaran** [comment]
> But, what about mobx?

Hi @thinkpadder1 👋 

HyperApp's is influenced by the Elm Architecture, hence the suitable comparison to Redux. In HyperApp the model is theoretically immutable, components are stateless, pure functions and actions are a mixture of reducers (old state + data -> new state out) and effects (reactions in Mobx).

Mobx seems to favor object-oriented programming, using classes/prototypes and stateful components. Use of observables, decorators and TypeScript is also popular.

At a glance, Mobx seems entirely different to HyperApp's model, so I have no idea how it could be used with Mobx. I'd be interested in hearing your ideas, though, of how the API would work and look like.

**Comment by jbucaran** [comment]
I looked at their examples, found a simple countdown timer example and made the same thing in HyperApp. Let's see how they compare:

<details><summary>Countdown Timer: <strong>HyperApp</strong></summary>

```html
<script src="https://unpkg.com/hyperapp"></script>
```

```jsx

```

[View online]()
</details>



<details><summary>Countdown Timer: <strong>React+Mobx</strong></summary>

```html
<script src="https://cdn.jsdelivr.net/g/react@15.4.2(react.min.js+react-dom.min.js),lodash@4.17.4,momentjs@2.14.1"></script>
<script src="https://unpkg.com/mobx@3.0.1/lib/mobx.umd.min.js"></script>
<script src="https://unpkg.com/mobx-react@4.1.0/index.js"></script>
<script src="https://unpkg.com/mobx-react-devtools"></script>
```


```jsx
mobx.useStrict(true);

var countdownTimerFactory = function(durationMilliseconds, options) {
    var intervalID;

    var timer = {
        id: _.uniqueId("countdownTimer_"),
        originalMilliseconds: durationMilliseconds
    };

    var settings = _.assign(
        {
            interval: 10,
            runTime: 0,
            resetOnComplete: true
        },
        options
    );

    mobx.extendObservable(timer, {
        durationAsMilliseconds: timer.originalMilliseconds,
        isTimerRunning: false,
        get isComplete() {
            return timer.durationAsMilliseconds <= 0;
        },
        get display() {
            return _.padStart(timer.minutesRemaining, 2, 0) +
                " : " +
                _.padStart(timer.secondsRemaining, 2, 0);
        },
        get durationAsDate() {
            return new Date(timer.durationAsMilliseconds);
        },
        get millisecondsRemaining() {
            return _.round(timer.durationAsDate.getUTCMilliseconds(), 2);
        },
        get secondsRemaining() {
            return timer.durationAsDate.getUTCSeconds();
        },
        get minutesRemaining() {
            return timer.durationAsDate.getUTCMinutes();
        },
        get hoursRemaining() {
            return timer.durationAsDate.getUTCHours();
        },
        get percentageComplete() {
            return 100 -
                _.round(
                    timer.durationAsMilliseconds /
                        timer.originalMilliseconds *
                        100,
                    2
                );
        },
        startTimer: mobx.action("startTimer", function() {
            timer.isTimerRunning = true;
        }),
        stopTimer: mobx.action("stopTimer", function() {
            timer.isTimerRunning = false;
        }),
        reset: mobx.action("resetTimer", function() {
            timer.stopTimer();
            timer.durationAsMilliseconds = timer.originalMilliseconds;
        })
    });

    mobx.autorun("countDownTimer", function() {
        if (timer.isTimerRunning) {
            intervalID = window.setInterval(
                function() {
                    mobx.runInAction("timer tick", function() {
                        timer.durationAsMilliseconds -= settings.interval;
                        if (timer.isComplete) {
                            timer.isTimerRunning = false;
                        }
                    });
                },
                settings.interval
            );
        } else if (intervalID) {
            if (settings.resetOnComplete && timer.isComplete) {
                timer.reset();
            }
            window.clearInterval(intervalID);
        }
    });

    return timer;
};

var Main = mobxReact.observer(function(props) {
    var timer = props.timer;
    return React.DOM.div(
        null,
        React.createElement(mobxDevtools.default),
        React.createElement(timerWithBar, { timer: timer }),
        React.createElement(timerControl, { timer: timer }, timerControl)
    );
});

var timerControl = mobxReact.observer(
    React.createClass({
        render: function() {
            var timer = this.props.timer;
            var button;

            if (timer.isTimerRunning) {
                button = React.DOM.button(
                    { onClick: this.handlePause },
                    "pause"
                );
            } else {
                button = React.DOM.button(
                    { onClick: this.handleStart },
                    "start"
                );
            }

            return React.DOM.div(null, button);
        },
        handleStart: function(event) {
            event.preventDefault();
            this.props.timer.startTimer();
        },
        handlePause: function(event) {
            event.preventDefault();
            this.props.timer.stopTimer();
        }
    })
);

var timerWithBar = mobxReact.observer(function(props) {
    var timer = props.timer;
    return React.DOM.div(
        { className: "active-blind" },
        React.createElement(timerPercentageCompleteRenderer, { timer: timer }),
        React.createElement(timerRenderer, { timer: timer })
    );
});

var timerPercentageCompleteRenderer = mobxReact.observer(function(props) {
    return React.DOM.div(
        { className: "progress-bar-container" },
        React.DOM.div({
            className: "progress-bar",
            style: { width: props.timer.percentageComplete + "%" }
        })
    );
});

var timerRenderer = mobxReact.observer(function(props) {
    var timer = props.timer;
    return React.DOM.div(null, timer.display);
});

var timer = countdownTimerFactory(10000);
ReactDOM.render(
    React.createElement(Main, {
        timer: timer
    }),
    document.getElementById("mount")
);

```

[View online](https://jsfiddle.net/gh/get/library/pure/mattruby/mobx-examples/tree/master/react-examples/60-countdown-timer)
</details>

What can we take from the above?

* Mobx has powerful ecosystem and devtools, it's flexible; it does not force you to think in terms of immutable state and reducers. 

* Mobx can feel frameworky, it's also pretty large, react bindings + mobx is like **64kb** minified, and 20kb gzipped. Add React, router, fetch and friends on top and you have 1/10 of a megabyte or more. Now you are ready to start writing _your app_!

* HyperApp's is still in its infancy, the [User Guide](https://www.gitbook.com/book/hyperapp/hyperapp) is still a work in progress and we need more people to try it out and make cool stuff with it

* HyperApp applications are functional out of the box, that's a selling point, but also presents a learning curve to newcomers used to traditional imperative / object-oriented programming (I know that Mobx is better than that, but I think they belong to the same category).

**Comment by tunnckoCore** [comment]
Damn, i'm tired and lazy, so i'll paste a pic of what i wrote while i was in lecture in university, haha, so hope you get it and let me know if there is anything to clarify.

![img_20170222_163428](https://cloud.githubusercontent.com/assets/5038030/23216010/1c8c45b8-f91d-11e6-99ea-4e9bde6e30c0.jpg)
![img_20170222_163414](https://cloud.githubusercontent.com/assets/5038030/23216011/1c8ecdc4-f91d-11e6-8bb1-aa99b7f021b3.jpg)

**Comment by thinkpadder1** [comment]
The timer example is much too simple. Mobx was originally made because the creator was working on a project with a lot of domain models (more than 500 for his application alone). Imagine trying to create that in Redux/Elm and with functional programming only.

[Here is an example](https://mobx.js.org/best/store.html) with a Todo Store.

**Comment by jbucaran** [comment]
@thinkpadder1 Sure. That was just one of the examples in Mobx examples repository.

It seems the link you've posted links to mostly documentation, so it's kind of hard to say anything or make any hard comparisons. I'd rather work on actual examples.

Do you have an example where Mobx really shines? I'd love to code it using HyperApp and see where that takes me.

And I agree with you, a countdown timer is much too simple, but that also begs the question, how come Mobx's implementation of it is so complex?

**Comment by jbucaran** [comment]
@thinkpadder1 Mobx is on the left. Both programs are doing the same thing. 

* HyperApp's example adds a reset button too, and has no dependencies; Mobx's has no reset button and includes lodash and moment.js as dependencies.

![hvsmobx](https://cloud.githubusercontent.com/assets/56996/23219004/334ccd2e-f961-11e6-957f-6e80e1413e77.png)

**Comment by tunnckoCore** [comment]
And why you deleted my comment? It is perfectly readable and is seen that Mobx absolutely doesn't fit well in elm-like project. It's not just it does fit, it totally can be compared - size and overhead at least.

**Comment by thinkpadder1** [comment]
@jbucaran That is literally the most convoluted example I have ever come across. No one writes their MobX applications like that. I can rewrite it if you would like.

Here is a full application:
- [Spectacle Editor](https://formidable.com/open-source/spectacle-editor/)
- [spectacle-editor source code](https://github.com/FormidableLabs/spectacle-editor)
- [Why we chose MobX over Redux for Spectacle Editor](https://formidable.com/blog/2016/06/02/why-we-chose-mobx-over-redux-for-spectacle-editor/)

**Comment by jbucaran** [comment]
@thinkpadder1 The example was taken directly from mobx's org, from their [mobx-examples](https://github.com/mobxjs/mobx-examples/tree/master/react-examples/60-countdown-timer) repo.

**Comment by thinkpadder1** [comment]
@jbucaran All those examples are written in ES5 and without JSX... He was just showing that you could do it without ES6+, decorators, typescript etc.

**Comment by jbucaran** [comment]
@thinkpadder1 Thanks for the links, unfortunately, that's not what I asked for. I'm still agnostic about Mobx's superiority over Redux or TEA (The Elm Architecture).

I think the correct way to debate here, is not with words, but code. I'll ask again, please provide an example where Mobx shines so that I can try to implement the HyperApp's version and learn HyperApp's architecture's choice shortcomings.

> All those examples are written in ES5 and without JSX... He was just showing that you could do it without ES6+, decorators, typescript etc.

Sure! That'd certainly improve things a bit, 30%? Not sure. Still, I think it's obvious that example was a poor choice to represent Mobx.

**Comment by thinkpadder1** [comment]
@jbucaran I thought [Spectacle Editor](https://github.com/FormidableLabs/spectacle-editor) was a good example, because of the amount of different models and how they react to one another.

How about this [remote control app](https://github.com/GPMDP/google-play-music-desktop-remote) for Google Play Desktop? It has a lot of side effects and inputs.

**Comment by jbucaran** [comment]
@thinkpadder1 Thanks, that's better than docs. I was hoping for some more down to earth examples. Something vanilla like [HyperApp's Examples](https://hyperapp.gomix.me/).

Mobx's org still has other examples that I didn't have time to look at, so I'll start by rewriting some of those to HyperApp. Maybe I'll find something.

For example, while writing the countdown timer, I noticed a few things which led to creating a few issues today:

* https://github.com/hyperapp/hyperapp/issues/105
* https://github.com/hyperapp/hyperapp/issues/106

The bottom line, all this tinkering is good, we can learn from Mobx too.

**Comment by thinkpadder1** [comment]
@jbucaran Sorry, all the smaller examples I have found were just too simple (not what you would use MobX for). 😢
There is also this [contacts list](https://github.com/mobxjs/mobx-contacts-list) application which is much simpler than the Google Play one.

**Comment by jbucaran** [comment]
@thinkpadder1 Great! The contacts list one looks simple (in scope), so I could probably rewrite it to HyperApp in a couple of days.

Thanks for bearing with me 🙇.

**Comment by jbucaran** [comment]
More examples are coming! In the meantime we should close here since hyperapp will not be adding mobx support anytime soon.
